### ISSUE TITLE
defaultDarkTheme bug and filtering for themes 

### ISSUE THREAD

**Issue by dwertheimer** [NotePlan/plugins#312]
Per discussion: https://discord.com/channels/763107030223290449/989752996583858217/989889481706471476

1) Add `Editor.themeName` property to get current theme. It would also be helpful to know if that theme is a light or dark and if we are in light or dark mode. I guess it could be `Editor.theme` which would return the whole object, and you could read the `name` and `style` field. But you still wouldn't know if the system is in light/dark mode.

2) Change how DataStore.preference() is set for themes. Default Light/Dark should not be overwritten just because it's day and you switched to a dark theme

3) Because we can't load and parse the theme files (esp the system ones), please add a filtering function to the availableThemes list. e.g. `Editor.availableThemesByStyle("Dark")`

**Comment by dwertheimer** [comment]
#2 is high priority
#1 and #3 are enhancement

**Comment by EduardMe** [comment]
Can I add a variable to the setTheme function:

`func setTheme(_ name: String, _ shouldSave: Bool)` = By default this would be `false` due to the way JavaScript works.

So you set the theme through the API, but if the user restarts or changes to dark mode, then light mode, etc, the OLD theme will re-appear, since it's not saved.

Did I get this right?

**Comment by dwertheimer** [comment]
I am mostly going to defer to @jgclark who is carrying this banner right now. But I think we are asking for something different than you described above.
1) We need a way to get the details of the currently active theme. Perhaps doing it in two parts:

a) Editor.activeTheme: string // returns the string theme name of the current theme
b) Editor.getThemeByName(themeName:string): Object/json // returns the entire JSON for the theme as an object (so @jgclark can process it and turn it into CSS) -- needs to return either built-in or user-created themes
c) Editor.isInDarkMode: boolean // tells whether the System is in light or dark mode

...I think the above would satisfy @jgclark's need for being able to load the theme details to make the popup consistent theme-wise with the running theme. Ultimately, it feels to me like this should be default behavior for the popup code, but @jgclark can prototype this functionality for you.

But so we don't lose these other things, which I still think are important for people who want to easily switch between themes using a quick switcher plugin:

2) Editor.availableThemesByStyle("Dark") - expand the existing theme list API to allow the filtering of the list to dark/light

3) I also think we need to change how DataStore.preference.defaultLightTheme, etc. is set for themes. Your DEFAULT Light/Dark should not be overwritten just because it's day and you switched to a dark theme. I quite often switch between themes (e.g. to go to a Monsopace theme, and this *does not* mean I want to change my default dark or light theme)

**Comment by jgclark** [comment]
I think @dwertheimer's analysis is right, and that most of these are requests aren't (fully) met yet.
In fact, @EduardMe, I think only 1a) is done with your reported addition to build~846 yesterday.
Thanks.

**Comment by EduardMe** [comment]
Yes, I'm aware, I just wanted to confirm #2 (of the original requests) with this comment (I should have been more clear):

> Change how DataStore.preference() is set for themes. Default Light/Dark should not be overwritten just because it's day and you switched to a dark theme

In the latest reply it's #3

> I also think we need to change how DataStore.preference.defaultLightTheme, etc. is set for themes. Your DEFAULT Light/Dark should not be overwritten just because it's day and you switched to a dark theme. I quite often switch between themes (e.g. to go to a Monsopace theme, and this does not mean I want to change my default dark or light theme)

I did add a `shouldSave` parameter to the `setTheme` function, so by default it won't save it. I'm just not sure if this is exactly what is required. Would this work?

**Comment by dwertheimer** [comment]
@EduardMe People have asked in the Theme Chooser plugin for me to make it possible to set dark and light default theme, no matter what time of day it is and what mode their computer is in. And switching to other themes during the day should not change what their "default" light and dark theme are. Can you pls tell me how to do that?

**Comment by EduardMe** [comment]
Ok, so switching to other themes without overwriting the currently saved one would work with `setTheme(themeName, false)` (default, you don't need to use `false` here). Then the theme should be changed, but not saved. So upon restart it goes back to default.

To save a theme, use `true`, but right now you can't tell NotePlan if it should save it for light or dark mode. That could be a third parameter or we change the `shouldSave` parameter to `mode` and if it's undefined, it will not save it. If it's `dark`, it will save it for dark. Or still use the `shouldSave` parameter? Then it would do this automatically.

**Comment by EduardMe** [comment]
I think I got everything added or extended now, reflected in this documentation (and available all together in the next alpha/beta update):

```javascript
Editor.

/**
* Note: Available from NotePlan v3.6.2+
* Change the current theme. Get all available theme names using `.availableThemes`. Custom themes are also supported. Use the filename in this case. This will not save the theme, it will just set it. If the system switched from light to dark mode, NotePlan will automatically load up the default (saved) theme for that mode.
* @param {String}
*/
.setTheme(name)  
  
/**
* Note: Available from NotePlan v3.6.2+
* Saves the default theme for a specific mode. It will not set the currently used theme, but just save it. So if you restart NotePlan or the system mode changes, it will use this saved default theme (if it's the same mode you saved it for). As mode use "dark", "light" or "auto" (uses the current system mode).
* @param {String}
* @param {String}
*/
.saveDefaultTheme(name, mode) 

/**
 * Note: Available from NotePlan v3.6.2+
 * Get all supported themes (including custom themes imported into the Theme folder) as an array of theme objects having four keys: "name" (name of the theme as seen in the json file), "mode" (dark or light), "filename" (original filename with "json" as extension) and "values" (parsed json file of the theme as object, so you can access every value like `Editor.availableThemes[0].values["styles"]["body"]["color"]`)
 *
 * Use together with `.setTheme(theme.name)`
 * @type [
 *   {
 *      name: String, 
 *      mode: String ("dark", "light"),
 *      filename: String,
 *      values: {...}
 *   }
 * ]
 */
.availableThemes
  
/**
 * Note: Available from NotePlan v3.6.2+
 * Get the current theme as an object with four keys: "name" (name of the theme as seen in the json file), "mode" (dark or light), "filename" (original filename with "json" as extension) and "values" (parsed json file of the theme as object, so you can access every value like `Editor.currentTheme.values["styles"]["body"]["color"]`). To get the name access Editor.currentTheme.name, for the mode, call Editor.currentTheme.mode, etc.
 * @type {
 *    name: String, 
 *    mode: String ("dark", "light"),
 *    filename: String,
 *    values: {...}
 * }
 */
.currentTheme
  
/**
 * Note: Available from NotePlan v3.6.2+
 * Get the current system mode, either "dark" or "light.
 * @type { String }
 */
.currentSystemMode
```

**Comment by dwertheimer** [comment]
@EduardMe Looks perfect to me. Thank you. @jgclark ?
